File: backend/app/schemas/admin_reports.py
"""Pydantic schemas for /admin/students endpoints (student reports + cohort)."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class StudentRow(BaseModel):
    """One row in the admin reports table."""

    model_config = ConfigDict(from_attributes=True)

    user_id: UUID
    full_name: str
    email: str
    level: str | None = None
    entry_score: float | None = None
    overall_progress_pct: float
    avg_quiz_score: float | None = None
    avg_coding_score: float | None = None
    last_activity_at: datetime | None = None
    last_location: str | None = None
    is_active: bool


class ModuleProgress(BaseModel):
    module_id: int
    module_title: str
    topics_total: int
    topics_completed: int
    progress_pct: float
    avg_quiz_score: float | None = None
    avg_coding_score: float | None = None


class QuizAttemptRow(BaseModel):
    attempt_id: int
    topic_id: int
    topic_title: str
    score: float
    is_passed: bool
    attempted_at: datetime


class CodingSubmissionRow(BaseModel):
    submission_id: int
    challenge_id: int
    challenge_title: str
    score: float
    submitted_at: datetime


class AchievementRow(BaseModel):
    achievement_id: int
    name: str
    badge_emoji: str
    earned_at: datetime


class LevelHistoryEntry(BaseModel):
    level: str
    score: float
    changed_at: datetime
    reason: str | None = None


class StudentDetail(BaseModel):
    user_id: UUID
    full_name: str
    email: str
    created_at: datetime
    is_active: bool
    level: str | None = None
    entry_score: float | None = None
    level_history: list[LevelHistoryEntry] = Field(default_factory=list)
    overall_progress_pct: float
    modules: list[ModuleProgress] = Field(default_factory=list)
    recent_quizzes: list[QuizAttemptRow] = Field(default_factory=list, max_length=10)
    recent_coding: list[CodingSubmissionRow] = Field(default_factory=list, max_length=10)
    chat_messages_count: int
    chat_last_at: datetime | None = None
    achievements_earned: list[AchievementRow] = Field(default_factory=list)
    total_time_seconds: int
    last_activity_at: datetime | None = None
    last_location: str | None = None


class AIReport(BaseModel):
    summary: str
    strengths: list[str]
    weaknesses: list[str]
    risk_level: str  # bajo | medio | alto
    risk_reason: str
    interventions: list[str]
    generated_at: datetime
    cached: bool


class CohortReportRequest(BaseModel):
    user_ids: list[UUID] = Field(min_length=2, max_length=15)


class CohortAIReport(BaseModel):
    narrative: str
    top_performers: list[str]
    needs_support: list[str]
    common_gaps: list[str]
    recommendations: list[str]
    generated_at: datetime
    cached: bool

File: frontend/src/types/adminReports.ts
import type { StudentLevel } from './assessment'

export interface StudentRow {
  user_id: string
  full_name: string
  email: string
  level: StudentLevel | null
  entry_score: number | null
  overall_progress_pct: number
  avg_quiz_score: number | null
  avg_coding_score: number | null
  last_activity_at: string | null
  last_location: string | null
  is_active: boolean
}

export interface ModuleProgress {
  module_id: number
  module_title: string
  topics_total: number
  topics_completed: number
  progress_pct: number
  avg_quiz_score: number | null
  avg_coding_score: number | null
}

export interface QuizAttemptRow {
  attempt_id: number
  topic_id: number
  topic_title: string
  score: number
  is_passed: boolean
  attempted_at: string
}

export interface CodingSubmissionRow {
  submission_id: number
  challenge_id: number
  challenge_title: string
  score: number
  submitted_at: string
}

export interface AchievementRow {
  achievement_id: number
  name: string
  badge_emoji: string
  earned_at: string
}

export interface LevelHistoryEntry {
  level: StudentLevel
  score: number
  changed_at: string
  reason: string | null
}

export interface StudentDetail {
  user_id: string
  full_name: string
  email: string
  created_at: string
  is_active: boolean
  level: StudentLevel | null
  entry_score: number | null
  level_history: LevelHistoryEntry[]
  overall_progress_pct: number
  modules: ModuleProgress[]
  recent_quizzes: QuizAttemptRow[]
  recent_coding: CodingSubmissionRow[]
  chat_messages_count: number
  chat_last_at: string | null
  achievements_earned: AchievementRow[]
  total_time_seconds: number
  last_activity_at: string | null
  last_location: string | null
}

export type RiskLevel = 'bajo' | 'medio' | 'alto'

export interface AIReport {
  summary: string
  strengths: string[]
  weaknesses: string[]
  risk_level: RiskLevel
  risk_reason: string
  interventions: string[]
  generated_at: string
  cached: boolean
}

export interface CohortAIReport {
  narrative: string
  top_performers: string[]
  needs_support: string[]
  common_gaps: string[]
  recommendations: string[]
  generated_at: string
  cached: boolean
}
